src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -394,13 +393,316 @@ pub trait TimestampProvider {
+        timeline_context: &TimelineContext,
+        oracle_read_ts: Option<Timestamp>,
+        real_time_recency_ts: Option<mz_repr::Timestamp>,
+        isolation_level: &IsolationLevel,
+        constraint_based: &ConstraintBasedTimestampSelection,
+    ) -> Result<
+        (
+            TimestampDetermination<mz_repr::Timestamp>,
+            ReadHolds<mz_repr::Timestamp>,
+        ),
+        AdapterError,
+    > {
+        // First, we acquire read holds that will ensure the queried collections
+        // stay queryable at the chosen timestamp.
+        let read_holds = self.acquire_read_holds(id_bundle);
+        let timeline = Self::get_timeline(timeline_context);
+
+        let since = self.least_valid_read(&read_holds);
+        let upper = self.least_valid_write(id_bundle);
+        let largest_not_in_advance_of_upper = Coordinator::largest_not_in_advance_of_upper(&upper);
+
+        let raw_determination = match constraint_based {
+            ConstraintBasedTimestampSelection::Disabled => self.determine_timestamp_classical(
+                session,
+                &read_holds,
+                id_bundle,
+                when,
+                oracle_read_ts,
+                compute_instance,
+                real_time_recency_ts,
+                isolation_level,
+                &timeline,
+                largest_not_in_advance_of_upper,
+                &since,
+            )?,
+            ConstraintBasedTimestampSelection::Enabled => self
+                .determine_timestamp_via_constraints(
+                    session,
+                    &read_holds,
+                    id_bundle,
+                    when,
+                    oracle_read_ts,
+                    compute_instance,
+                    real_time_recency_ts,
+                    isolation_level,
+                    &timeline,
+                    largest_not_in_advance_of_upper,
+                )?,
+            ConstraintBasedTimestampSelection::Verify => {
+                let classical_determination = self.determine_timestamp_classical(
+                    session,
+                    &read_holds,
+                    id_bundle,
+                    when,
+                    oracle_read_ts,
+                    compute_instance,
+                    real_time_recency_ts,
+                    isolation_level,
+                    &timeline,
+                    largest_not_in_advance_of_upper,
+                    &since,
+                )?;
+
+                let constraint_determination = self.determine_timestamp_via_constraints(
+                    session,
+                    &read_holds,
+                    id_bundle,
+                    when,
+                    oracle_read_ts,
+                    compute_instance,
+                    real_time_recency_ts,
+                    isolation_level,
+                    &timeline,
+                    largest_not_in_advance_of_upper,
+                )?;
```

Comment:
In `Verify`, we probably don't want to error the call if the constraints-based determination fails, rather log the error (instead of `?`).